netutils/webclient/webclient.c

```diff
@@ -674,225 +693,336 @@ int webclient_perform(FAR struct webclient_context *ctx)
     {
-      if (!strcmp(ws->scheme, "https") && tls_ops != NULL)
-        {
-          conn.tls = true;
-        }
-      else if (!strcmp(ws->scheme, "http"))
+      if (ws->state == WEBCLIENT_STATE_SOCKET)
         {
-          conn.tls = false;
-        }
-      else
-        {
-          nerr("ERROR: unsupported scheme: %s\n", ws->scheme);
-          free(ws);
-          return -ENOTSUP;
-        }
-
-      /* Re-initialize portions of the state structure that could have
-       * been left from the previous time through the loop and should not
-       * persist with the new connection.
-       */
-
-      ws->httpstatus = HTTPSTATUS_NONE;
-      ws->offset     = 0;
-      ws->datend     = 0;
-      ws->ndx        = 0;
-
-      if (conn.tls)
-        {
-          char port_str[sizeof("65535")];
-
-#if defined(CONFIG_WEBCLIENT_NET_LOCAL)
-          if (ctx->unix_socket_path != NULL)
+          if (!strcmp(ws->scheme, "https") && tls_ops != NULL)
+            {
+              conn->tls = true;
+            }
+          else if (!strcmp(ws->scheme, "http"))
             {
-              nerr("ERROR: TLS on AF_LOCAL socket is not implemented\n");
+              conn->tls = false;
+            }
+          else
+            {
+              nerr("ERROR: unsupported scheme: %s\n", ws->scheme);
               free(ws);
               return -ENOTSUP;
             }
-#endif
 
-          snprintf(port_str, sizeof(port_str), "%u", ws->port);
-          ret = tls_ops->connect(tls_ctx, ws->hostname, port_str,
-                                 CONFIG_WEBCLIENT_TIMEOUT, &conn.tls_conn);
-        }
-      else
-        {
-#if defined(CONFIG_WEBCLIENT_NET_LOCAL)
-          struct sockaddr_un server_un;
-#endif
-          struct sockaddr_in server_in;
-          int domain;
-          const struct sockaddr *server_address;
-          socklen_t server_address_len;
+          /* Re-initialize portions of the state structure that could have
+           * been left from the previous time through the loop and should not
+           * persist with the new connection.
+           */
 
-#if defined(CONFIG_WEBCLIENT_NET_LOCAL)
-          if (ctx->unix_socket_path != NULL)
-            {
-              domain = PF_LOCAL;
+          ws->httpstatus = HTTPSTATUS_NONE;
+          ws->offset     = 0;
+          ws->datend     = 0;
+          ws->ndx        = 0;
+          ws->redirected = 0;
 
-              memset(&server_un, 0, sizeof(server_un));
-              server_un.sun_family = AF_LOCAL;
-              strncpy(server_un.sun_path, ctx->unix_socket_path,
-                      sizeof(server_un.sun_path));
-#if !defined(__NuttX__) && !defined(__linux__)
-              server_un.sun_len = SUN_LEN(&server_un);
+          if (conn->tls)
+            {
+#if defined(CONFIG_WEBCLIENT_NET_LOCAL)
+              if (ctx->unix_socket_path != NULL)
+                {
+                  nerr("ERROR: TLS on AF_LOCAL socket is not implemented\n");
+                  free(ws);
```

Comment:
```suggestion
                  free(ws);
                  ctx->ws = NULL;
```
To avoid use-after-free errors, the freed pointer should be nullified.